scheduler/pkg/envoy/processor/incremental.go

```diff
@@ -185,57 +173,17 @@ func (p *IncrementalProcessor) handleModelEvents(event coordinator.ModelEventMsg
 	}()
 }
 
-func (p *IncrementalProcessor) init() error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	err := p.xdsCache.SetupTLS()
-	if err != nil {
-		return err
-	}
-	p.xdsCache.AddPermanentListeners()
-	p.xdsCache.AddPermanentClusters()
-	return nil
-}
-
-// newSnapshotVersion increments the current snapshotVersion
-// and returns as a string.
-func (p *IncrementalProcessor) newSnapshotVersion() string {
-	// Reset the snapshotVersion if it ever hits max size.
-	if p.snapshotVersion == math.MaxInt64 {
-		p.snapshotVersion = 0
-	}
-
-	// Increment the snapshot version & return as string.
-	p.snapshotVersion++
-	return strconv.FormatInt(p.snapshotVersion, 10)
-}
-
 func (p *IncrementalProcessor) updateEnvoy() error {
-	logger := p.logger.WithField("func", "updateEnvoy")
-	// Create the snapshot that we'll serve to Envoy
-	snapshot, err := cache.NewSnapshot(
-		p.newSnapshotVersion(), // version
-		map[rsrc.Type][]types.Resource{
-			rsrc.ClusterType:  p.xdsCache.ClusterContents(),  // clusters
-			rsrc.RouteType:    p.xdsCache.RouteContents(),    // routes
-			rsrc.ListenerType: p.xdsCache.ListenerContents(), // listeners
-			rsrc.SecretType:   p.xdsCache.SecretContents(),   // Secrets
-		})
+	// sequencing is important here - send clusters before updating routes and removing old clusters
+	err := p.xdsCache.AddClusters()
 	if err != nil {
 		return err
 	}
-
-	if err := snapshot.Consistent(); err != nil {
-		return err
-	}
-	logger.Debugf("will serve snapshot %+v", snapshot)
-
-	// Add the snapshot to the cache
-	if err := p.cache.SetSnapshot(context.Background(), p.nodeID, snapshot); err != nil {
+	err = p.xdsCache.UpdateRoutes(p.nodeID)
+	if err != nil {
 		return err
 	}
-
-	return nil
+	return p.xdsCache.RemoveClusters()
```

Comment:
This is the crux of the change, as the sequence is now: add new clusters, point routes to new clusters, remove old clusters. It's required because incremental clears out a route before remaking it (whether it's a model, or pipeline route) and this intermediate state shouldn't be sent to envoy.